proxy/__init__.py

```diff
@@ -0,0 +1,201 @@
+"""
+from __future__ import annotations
+from typing import Any
+import re
+import json
+
+import requests
+
+from app import model_cache
+from data.cache import cache_key
+from data.database import ProxyCacheConfig
+from digest.digest_tools import sha256_digest
+
+
+WWW_AUTHENTICATE_REGEX = re.compile(r'(\w+)[=] ?"?([^",]+)"?')
+TOKEN_VALIDITY_LIFETIME_S = 60 * 60  # 1 hour, in seconds - Quay's default
+TOKEN_RENEWAL_THRESHOLD = 10  # interval (in seconds) when to renew auth token
+
+
+REGISTRY_URLS = {"docker.io": "registry-1.docker.io"}
+
+
+class UpstreamRegistryError(Exception):
+    def __init__(self, detail):
+        msg = f"Error requesting upstream registry: {detail}"
+        super().__init__(msg)
+
+
+def parse_www_auth(value: str) -> dict[str, str]:
+    """
+    Parses WWW-Authenticate parameters and returns a dict of key=val.
+    See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate
+    for details.
+    This parser is *not* fully compliant with RFC 7235, notably it does not support
+    multiple challenges.
+    """
+    scheme = value.split(" ", 1)[0]
+    matches = WWW_AUTHENTICATE_REGEX.findall(value)
+    parsed = dict(matches)
+    if scheme:
+        parsed["scheme"] = scheme
+    return parsed
+
+
+class Proxy:
+    def __init__(self, config: ProxyCacheConfig, repository: str):
+        self._config = config
+
+        hostname = REGISTRY_URLS.get(
+            config.upstream_registry_hostname,
+            config.upstream_registry_hostname,
+        )
+        url = f"https://{hostname}"
+        if config.insecure:
+            url = f"http://{hostname}"
+
+        self.base_url = url
+        self._session = requests.Session()
+        self._repo = repository
+        self._authorize(self._credentials())
+
+    def get_manifest(self, image_ref: str, media_type: str | None = None) -> tuple[str, str | None]:
+        headers = {}
+        if media_type is not None:
+            headers["Accept"] = media_type
+
+        url = f"{self.base_url}/v2/{self._repo}/manifests/{image_ref}"
+        resp = self.get(url, headers=headers)
+        raw_manifest = resp.content
+        content_type = resp.headers.get("content-type")
+        return raw_manifest, content_type
+
+    def manifest_exists(self, image_ref: str, media_type: str | None = None) -> str:
+        url = f"{self.base_url}/v2/{self._repo}/manifests/{image_ref}"
+        headers = {}
+        if media_type is not None:
+            headers["Accept"] = media_type
+        resp = self.head(url, headers=headers, allow_redirects=True)
+        return resp.headers.get("docker-content-digest")
+
+    def get_blob(self, digest: str):
+        url = f"{self.base_url}/v2/{self._repo}/blobs/{digest}"
+        resp = self.get(
+            url,
+            allow_redirects=True,
+        )
+        size = resp.headers.get("content-length")
+        content_digest = sha256_digest(resp.content)
```

Comment:
Will it load the entire blob in memory? We should never do it.

Reply:
yeah it will :/ 
is this something we want to address in this PR? I have a feeling that not doing so won't exactly be straight forward.

Reply:
Discussed on slack and although this is a release blocker, it'll be done in a follow-up PR. See https://issues.redhat.com/browse/PROJQUAY-3459